### ISSUE TITLE
Add a new grouped filter in OmnibusNoiseFilter

### ISSUE THREAD

**Issue by wenqiang-gu** [WireCell/wire-cell-toolkit#327]
While Xuyang is developing noise filters for ProtoDUNE HD, a new type of "nosie" occurs in half of FEMB with large negative pulses. Currently, the coherent noise filter in PDHD bundles all 40 U channels from a FEMB in a group. And similarly, all 40 V channels from a FEMB, all 48 W channels from a FEMB. Therefore, these two grouping policies are different. 

Here is how the coherent noise filter is implemented in OmnibusNoiseFilter:

https://github.com/WireCell/wire-cell-toolkit/blob/1a00d641b90d500238c4696bc204f45905c2a454/sigproc/src/OmnibusNoiseFilter.cxx#L187

I would like to add a different filter, e.g.,
```
for (auto group: m_noisedb->partial_femb() ) {

...
        for (auto filter : m_partial_femb) {
            auto masks = filter->apply(chgrp);
            Waveform::merge(cmm, masks, m_maskmap);
        }
...
}
```
Would this proposal be align with the overall design of wire-cell-toolkit?

**Comment by wenqiang-gu** [comment]
@brettviren Could you comment on this proposal?

Also, could you add Xuyang (GitHub username: Ningclover)as a core developer in GitHub? She needs to develop this filter.

**Comment by brettviren** [comment]
Hi @wenqiang-gu 

Almost, but I think not quite.  We should not see a string like "femb" inside WCT C++ as that is a detector-specific idea.

To refactor the problem and your proposal more generically: right now WCT accepts only a single association between a single "set of channel groups" (`{CG}`) and  a single "set of group filters" (`{GF}`):

```
 {CG} <--> {GF}
```
What we want is to extend this to allow multiple associations:
```
 {CG1} <--> {GF1}
 {CG2} <--> {GF2}
 {CG3} <--> {GF3}
```

Just to define a label, I call this: "multi group coherent channels" (MGCC). We can then say that WCT currently only supports "single group coherent channels" (SGCC).

To add support for "MGCC" we must keep in mind backwards compatibility.  Also, I think it is best to minimize the footprint (amount of code change) of the extension.

So, I suggest we do the following:

1. **Ignore** `OmniChannelNoiseDB` and focus only on `OmnibusNoiseFilter`.
2. **Remove** the current "SGCC loop" in `OmnibusNoiseFilter`:
``` c++
    for (auto group : m_noisedb->coherent_channels()) {
       ...
```
3. **Replace** the SGCC loop with a new MGCC loop.  It will be similar but instead iterates over MGCC information (I'll describe).  This will need a data member on `OmnibusNoiseFilter` like `std::vector<MGCC> m_multigroup_coherent`) to provide ready-to-use representation of MGCC.  The new MGCC loop would then look something like:
``` c++
    for (const auto& [mgcc] : m_multigroup_coherent)
        for (auto ch : mgcc.channels) {
            // fill chgrp as in the SGCC loop
        }
        for (auto filter : mgcc.filters) {
            // apply filter as in the SGCC loop
            auto masks = filter->apply(chgrp);
            Waveform::merge(cmm, masks, m_maskmap);
        }
        // ... etc as the SGCC loop
```
4. **Introduce** a new configuration attribute to `OmnibusNoiseFilter` that would allow Jsonnet like:

``` jsonnet
local filter1 = /* usual IChannelFilter config */;
local filter2 = /* usual IChannelFilter config */;
type: 'OmnibusNoiseFilter',
data: {
  multigroup_coherent: [
    {
      channels: [list of channel IDs],
      filters: [wc.tn(filter1), wc.tn(filter2)]
     },
     { ... },
     { ... },
     { ... },
  ],
}, // ...
```
5. **Define** new little `struct` called `MGCC` inside `OmnibusNoiseFilter` like:
``` c++
struct MGCC {
  std::vector<int> channels;
  std::vector<IChannelFilter::pointer> filters;
};
```
6. **Construct** the `m_multigroup_coherent` in `OmnibusNoiseFilter::configure()` from two pieces of information.  First, fill it with the SGCC information from `m_noisedb->coherent_channels()` and `m_grouped`.  Second, fill it from information from the `multigroup_coherent` configuration attribute.


Note, I have intentionally excluded `OmniChannelNoiseDB` because I believe MGCC information is only required inside `OmnibusNoiseFitler`.  If that is wrong, we may need to extend both `OmnibusNoiseFitler` and `OmniChannelNoiseDB`.

Also, I wrote the above so that MGCC support is a superset that covers SGCC.  That is, we may - in principle - (re)move existing SGCC configuration and have the same functionality expressed as MGCC.  It is possible I missed something that would not allow this design to be a superset of SGCC functionality.  If so we should fix the design to allow the superset.  And, I don't mean to imply we **must** remove SGCC config, just that it should be possible.  Existing experiments may not want to change.  But, new configuration can be made simpler.

Sorry this is long, but I hope it makes sense.  We can of course discuss.

(And, I'll invite Xuyang).

**Comment by wenqiang-gu** [comment]
It sounds good to me. In fact, I think we can even totally remove `OmnibusNoiseDB` from OmnibusNoiseFilter as the current usage of `OmnibusNoiseDB` is just: `m_noisedb->bad_channels()` and `m_noisedb->cohenrent_channels()`.  Once we implement MGCC, we can pass those channels from the jsonnet level, e.g.,

```
type: 'OmnibusNoiseFilter',
data: {
  multigroup_coherent: [
    {
      channels: chndb.groups;
      filters: [wc.tn(filter1), wc.tn(filter2)]
     },
    {
      channels: chndb.bad;
      filters: [wc.tn(filter1), wc.tn(filter2)]
     },
     { ... },
     { ... },
  ],
}, // ...
```

**Comment by wenqiang-gu** [comment]
> I think we can even totally remove `OmnibusNoiseDB` from OmnibusNoiseFilter

Never mind, I recall you mentioned backwards compatibility. 

I am working on the implementation now. PS, can we name the struct MGCF (multi-group channel filters)? "coherent" seems a bit jargon-heavy.

**Comment by wenqiang-gu** [comment]
Hi @brettviren , I submitted my PR: https://github.com/WireCell/wire-cell-toolkit/pull/328

It's implemented by following what you proposed except that:
1) MGCC -> MGCF
2) use vector<vector<int>> in MGCF
```c++
            struct MGCF {
                std::vector<IChannelNoiseDatabase::channel_group_t> channelgroups;
                std::vector<IChannelFilter::pointer> filters;
            };
```

**Comment by brettviren** [comment]
Looks good, @wenqiang-gu .  And, yes `v<v<i>>` is correct and the naming is good.

I reviewed the PR.  It looks good.  Feel free to merge if/when you are ready.

---

An aside about removing CNDB from NF.

I do like your idea about removing CNDB entirely from NF as it would simplify
configuration substantially.  But of course, it would be a "breaking change".

If WCT releases also included all experiments' config or otherwise we would
release experiments' configuration tied to C++ releases, we could make such
changes easily.  

But since we do not have this level of control on the configs we must introduce breaking changes by following some kind of "deprecation procedure".  This procedure would span several releases amd communicate our intention to the users.  The procedure would have at least two stages.

1. Immediately we may add a `log.warn("Use of CHND is deprecated and will be removed in the future")` to print by NF if it is configured to use a CNDB object.  Besides the warning, the code works as normal (eg, as in your PR).
2. We pick a future "major release" of WCT to actually remove support for NF to use a CHND.  The release note details the "breaking change" and give users guidance how to upgrade their configuration.  To be friendly, we keep the check for any CNDB being configured but change it from warning to critical error: `log.critical("Use of CNDB is obsolete, see notes for release X.Y.Z'); raise<ValueError>("unexpected configuration: CNDB")`.  This will catch new code using old config.  To be super friendly, maybe the release includes a script to help the user to convert old config to new config.

In any case, we should not couple this possible cleanup with your PR here.  We can discuss to see if the cleanup of removing CNDB is worth the effort.

**Comment by wenqiang-gu** [comment]
Thank you for the comment. I added a warning for later reference:
```c++
log->warn("Use of CHND is deprecated and will be removed in the future.”);
```

**Comment by wenqiang-gu** [comment]
Recently, Herb reported a segmentation fault in uboone reco1 stage in MCC10. It turns out the problem is from this new feature of MGCF, and MicroBooNE's usage of [wclsMultiChannelNoiseDB](https://github.com/WireCell/wire-cell-toolkit/blob/398de01a58a6cc59ac64676f2805f5a579bcfd05/cfg/pgrapher/experiment/uboone/chndb.jsonnet#L36) is not consistent with this feature's implementation.

TLDR: cannot get `noisedb->coherent_channels()` in the `OmnibusNoiseFilter::configure()` when wclsMultiChannelNoiseDB is used.


---- More details ----
In the current `OmnibusNoiseFilter::configure()`, one tries to get `noisedb->coherent_channels()` ([here](https://github.com/WireCell/wire-cell-toolkit/blob/398de01a58a6cc59ac64676f2805f5a579bcfd05/sigproc/src/OmnibusNoiseFilter.cxx#L73)). However, in the MicroBooNE case, `coherent_channels()` does not return the expected result because it's not set in wclsMultiChannelNoiseDB. Instead, the standard noisedb is available in each `rule` of wclsMultiChannelNoiseDB.

As in the old version, `coherent_channels()` is called in the function of `OmnibusNoiseFilter::operator()`, where the `noisedb->coherent_channels()` would return expected reuslt.

**Comment by wenqiang-gu** [comment]
To avoid this issue in MicroBooNE's MCC10 production, I suggest two options:

1) stick with wirecel v0.27.1 since MicroBooNE does not need new features

2) introduce some minor changes to MicroBooNE's configuration:
```
// nf.jsonnet
            grouped_filters: [
                // wc.tn(grouped), // don't use the old interface
            ],
            multigroup_chanfilters: [ // use the new interface
                {
                  channelgroups: chndbobj.data.groups,
                  filters: [wc.tn(grouped)],
                }
            ],

// chndb.jsonnet
    wcls_multi: function(name="") {
        local bef = $.wcls("before"),
        ...
        type: "wclsMultiChannelNoiseDB",
        data: {
            groups: bef.data.groups, // introduce coherent channels from chndb-base
            rules: [...],
        },
    },
```

I am going to close this ticket but will discuss offline with stakeholders.
